firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java

```diff
@@ -153,29 +152,50 @@ private int indexOfOpenBracket() {
     }
   }
 
+  private class ReadJsonResult {
+    private String json;
+    private int byteCount;
+
+    ReadJsonResult(String json, int byteCount) {
+      this.json = json;
+      this.byteCount = byteCount;
+    }
+
+    String getJson() {
+      return json;
+    }
+
+    int getByteCount() {
+      return byteCount;
+    }
+  }
+
   /**
    * Reads from a specified position from the internal buffer, for a specified number of bytes,
    * pulling more data from the underlying stream if needed.
    *
-   * <p>Returns a string decoded from the read bytes.
+   * <p>Returns an object containing the Json string and its UTF8 byte count.
    */
-  private String readJsonString(int length) throws IOException {
+  private ReadJsonResult readJson(int length) throws IOException {
     StringBuilder json = new StringBuilder(length);
 
     int remaining = length;
+    int bytesRead = 0;
     while (remaining > 0) {
       if (buffer.remaining() == 0 && !pullMoreData()) {
         throw abort("Reached the end of bundle when more data was expected.");
       }
 
       int read = Math.min(remaining, buffer.remaining());
-      json.append(buffer, 0, read);
-      buffer.position(buffer.position() + read);
+      byte[] bytes = new byte[read];
+      buffer.get(bytes);
+      json.append(new String(bytes, charset));
```

Comment:
drive-by comment: This is dangerous because `bytes` could end in the middle of a multi-byte character. You could possibly use a class like `java.nio.charset.CharsetDecoder` to handle this properly. 

https://docs.oracle.com/javase/7/docs/api/java/nio/charset/CharsetEncoder.html

Reply:
Done.

Reply:
The change doesn't fix the issue. For example, consider the situation where the last byte of `bytes` specified to `new String(bytes, charset)` is the first byte of a multi-byte character, and the second byte of that character would be the first byte of the `bytes` in the next iteration; this would probably either fail decoding with an exception or would produce garbage.

Reply:
This is the bytes the bundle tells the SDK that it is a UTF8 string. If what you described happens, it means the bundle is invalid, and an exception is exactly what we want here.

Reply:
IIUC the code reads the bundle's bytes in "chunks". Wouldn't there be arbitrary breaks in the "chunks" such that a multibyte character could span two chunks?

Reply:
But now, what if `bytesToRead` > `BUFFER_CAPACITY`? IIUC, with the updated code, if the buffer fills up then `pullMoreData()` will return `false` and `readJson()` will incorrectly throw `IllegalArgumentException`.

I have another idea. What if you go back to the original implementation but change `StringBuilder` to `ByteArrayOutputStream`. You can then accumulate the bytes in the `ByteArrayOutputStream` and, once all bytes are read, you can decode the entire byte sequence, avoiding the multi-byte-character-spanning-a-chunk issue.

Reply:
Yeah, that is better.